File: frameworks/plugins/oppo/oppo.cpp
#include "oppo.h"

#if defined(CCLUA_BUILD_OPPO) && defined(CCLUA_OS_ANDROID)

USING_NS_CCLUA;
USING_NS_CCLUA_PLUGIN;

const char *oppo::JAVA_CLASS = "cclua/plugin/oppo/Oppo";
Callback oppo::_dispatcher;

void oppo::init()
{
    Jni::callStaticVoidMethod(JAVA_CLASS, "init");
}

void oppo::pay(const std::string &orderNo, const std::string &attach, int price, 
        const std::string &name, const std::string &desc, const std::string &url)
{
    Jni::callStaticVoidMethod(JAVA_CLASS, "pay", orderNo, attach, price, name, desc, url);
}

#endif

File: assets/src/cclua/plugin/oppo.lua
local util      = require "cclua.util"
local runtime   = require "cclua.runtime"

local oppo = runtime.load("cclua.plugin.oppo")
local oppo_pay = assert(oppo.pay, "pay")

local payCallback = nil

oppo.setDispatcher(function (event, data)
    print("oppo", event, util.dump(data))
    if event == "pay" then
        if not payCallback then
            print("oppo pay callback not set")
            return
        end
        if data.errcode == 1001 then
            payCallback("successful", data)
        elseif data.errcode == 1004 then
            payCallback("canceled", data)
        else
            payCallback("failed", data)
        end
        payCallback = nil
    end
end)
oppo.setDispatcher = false

function oppo.pay(order, callback)
    payCallback = assert(callback)
    oppo_pay(
        tostring(assert(order.order_no, "no order number")),
        tostring(assert(order.attach, "no attach")),
        tostring(assert(order.price, "no price")),
        tostring(assert(order.name, "no name")),
        tostring(assert(order.desc, "no desc")),
        tostring(assert(order.notify_url, "no notify url"))
    )
end

return oppo